### ISSUE TITLE
Question on Production Use-Cases

### ISSUE THREAD

**Issue by logan-markewich** [microsoft/i-Code#11]
Hello! 

Firstly, thanks for your fantastic efforts and research. The UDOP model is definitely one of the more interesting releases from the past few months.

Looking at the model code and reading the paper, I noticed you have decided to use a Seq2Seq type of format, very similar to T5. While this is a flexible type of model (as your paper proves!), I feel the largest downside is a lack of confidences for answers given.

For example, using LayoutLM1/2/3 for QA, I can easily filter out low-confidence answers, which makes production use-cases very feasible. With conditional generation, this is not as simple. In the past, I've tried to look at the confidences for each word generated, but that hasn't proved very useful.

Is there a particular method you suggest for filtering responses from UDOP or similar conditional generational models (i.e Donut, T5, etc.)? Or is this an area that might need more research in the future?

**Comment by zinengtang** [comment]
Generative models are searching answers from a very large space compared to discriminative models that only search from context or a limited label pools. But there are two cases of QA. 

In the case of extractive QA, i.e., answers are in the context, we can give generative model a prompt to tell that it is extractive. So, the model would know that it should obtain answer only from context and it can be more confident. This prompt needs to appear in pretraining of course.

But for QA with answers not in the context, we can only use generative models. It is exactly the nature of generative models to generate versatile outputs. For this part, it depends the tasks you are dealing with and the dataset trained on. What is the specific example you can give?

**Comment by logan-markewich** [comment]
Yes, I should have clarified my interest was in extractive QA, like the DocVQA dataset. When asking questions to a model trained on DocVQA, it can be very helpful to use the confidence to filter the answer and only show an end user the most confident predictions. Of course, this only works when predicting simple start/end positions of answers.

The paper doesn't specify exactly, but I'm assuming you used UDOP as a generative model when training/evaluating on DocVQA?

**Comment by zinengtang** [comment]
Yes it is generative model.
Usually we use beam search or greedy search and obtain the best sequence. The confidence here is joint probability of the searched words. I am not sure why finding the confidence is an issue.

**Comment by logan-markewich** [comment]
Finding the confidence is easy enough for sure! In the past, I've just found the confidence of the generated string is less helpful compared to the traditional start/end logits. Most recently I was using the Donut model after training on DocVQA. Often the confidences do not correlate with an answer being right or wrong. Instead, I had to rely on the model predicting an "answer not found" token, but even that was not as helpful.

Contrasted with traditional start/end position predictions, those confidences are an excellent indicator of a correct or wrong answer.

In any case, I am excited to see if this situation has improved with UDOP :)

**Comment by zinengtang** [comment]
Why in traditional start/end position predictions, confidences are excellent indicator? Do you have any experiments or stats to support this claim? For example, you can show that choosing the most confident answer is better than choosing from top5 in generative model.
>95 % does not mean anything absolutely but only relatively. >95% does not mean model is confident.

**Comment by logan-markewich** [comment]
Yes, 95% is relative. But if you don't use softmax, you can get a better indication of absolute confidence. So, maybe I shouldn't be using percentages here haha, it's more of a signal strength then.

I've been doing a lot of work with extractive QA. Unfortunately, it's all NDA and whatnot 🙄, but broadly speaking, I am extracting specific fields from documents, and each field has multiple questions that might lead to a desired answer. However, there is no guarantee each field is in the document.

For each field, one can determine an appropriate threshold, after testing on a large enough test set, so that the extracted answer is most likely correct. 

Usually, these confidence thresholds range from 20-60. But with my experiments with generative models like Donut, this process starts to fall apart. I wish I could share the data, but it seems others on the Donut github shared similar thoughts.

In any case, this is just my thoughts/experiences from a commercial perspective. I'm still very excited to take UDOP for a spin 💪🏻
